docs/proposals/kubelet-eviction.md

```diff
@@ -0,0 +1,219 @@
+its defined threshold.
+
+The eviction sequence works as follows:
+
+* for each monitoring interval, if eviction thresholds have been met
+ * find candidate pod
+ * fail the pod
+ * block until pod is terminated on node
+
+The `kubelet` will respect graceful termination periods for pods when
+doing eviction.  If a pod is not terminated within its graceful termination
+period because a container does not happen to die (i.e. processes stuck in
+disk IO for example), the `kubelet` may select an additional pod to fail.
+
+## Eviction Strategy
+
+The `kubelet` will implement a default eviction strategy oriented around
+the pod quality of service class.
+
+It will target pods that are the largest consumers of the starved compute
+resource relative to their scheduling request.  It ranks pods within a
+quality of service tier in the following order.
+
+* `BestEffort` pods that consume the most of the starved resource are failed
+first.
+* `Burstable` pods that consume the greatest amount of the starved resource
+relative to their request for that resource are killed first.  If no pod
+has exceeded its request, the strategy targets the largest consumer of the
+starved resource.
+* `Guaranteed` pods that consume the greatest amount of the starved resource
+relative to their request are killed first.  If no pod has exceeded its request,
+the strategy targets the largest consumer of the starved resource.
+
+Future iterations on this strategy would include the following:
+
+* target `BestEffort` pods that were most recently scheduled on the node
+* defer targeting pods that have data locality concerns, for example, its
+best to not evict a DB if it requires a local SSD
+* target pods with lower warm-up times before they are ready to accept traffic
+since they would be ready quicker on other nodes
+* etc.
+
+### Eviction best practices: DaemonSet
+
+It is never desired for a `kubelet` to evict a pod that was derived from
+a `DaemonSet` since the pod will immediately be recreated and rescheduled
+back to the same node.
+
+At the moment, the `kubelet` has no ability to distinguish a pod created
+from `DaemonSet` versus any other object.  If/when that information is
+available, the `kubelet` could pro-actively filter those pods from the
+candidate set of pods provided to the eviction strategy.
+
+In general, it should be strongly recommended that `DaemonSet` not
+create `BestEffort` pods to avoid being identified as a candidate pod
+for eviction.
+
+## Improve node stability for BestEffort pods
+
+### Feasibility checks during kubelet admission
+
+The `kubelet` will reject `BestEffort` pods if any of its associated
+eviction thresholds have been exceeded independent of the configured
+grace period.
+
+Let's assume the operator started the `kubelet` with the following:
+
+```
+--eviction-soft="memory.available<256Mi"
+--eviction-soft-grace-period="memory.available=30s"
+```
+
+If the `kubelet` sees that it has less than `256Mi` of memory available
```

Comment:
The scheduler will soon support usage based scheduling. Whatever logic that kubelet follows should be shared with the scheduler as well ideally. 
If the node is partitioned statically, this section effectively reduces the amount of `Available` resources right?